circ_ci_chunk.py

```diff
@@ -0,0 +1,295 @@
+    optimal_kappa1 = kappa1_chunk[idx_min]
+    optimal_kappa2 = kappa2_chunk[idx_min]
+    min_error = np.min(errors, axis=1)
+    np.save(f'./learned_data/errors_{task_idx}.npy', arr=errors)
+
+    print("Process ID: {}, mean shapes: {}, {}, min error {}".format(os.getpid(), mu1.shape, mu2.shape, min_error))
+
+    return (mean_indices, p_common, (optimal_kappa1, optimal_kappa2), min_error)
+
+def find_optimal_kappas():
+    """
+    Finds the optimal kappa values that minimize the error between model predictions and GAM data
+    by processing chunks of kappa combinations in parallel using multiprocessing.
+
+    This function prepares tasks for different combinations of mean indices, probabilities of a common cause (p_common),
+    and chunks of kappa values. It then uses a multiprocessing pool to process these tasks concurrently.
+    After processing, it collects and combines the results to find the kappa pairs that achieve the minimum error
+    for each mean index and p_common value.
+
+    Returns:
+        tuple: A tuple containing:
+            - optimal_kappa_pairs (dict): A dictionary mapping (mean index, p_common) tuples to optimal kappa pairs (kappa1, kappa2).
+            - min_error_for_idx_pc (dict): A dictionary mapping (mean index, p_common) tuples to the minimum error achieved.
+    """
+    tasks = []
+    print(f'Fitting for num_means={ut.shape}, data_shape={r_n.shape}')
+
+    # Adjust based on memory availability
+    chunk_size = 1000
+
+    total_kappa_combinations = len(kappa1_flat)
+    kappa_indices = np.arange(total_kappa_combinations)
+
+    num_chunks = (total_kappa_combinations + chunk_size - 1) // chunk_size
+    task_idx = 0
+    task_metadata = {}
+    for i, data_slice in enumerate(r_n):
+        for p_common in p_commons:
+            for chunk_idx in range(num_chunks):
+                start_idx = chunk_idx * chunk_size
+                end_idx = min((chunk_idx + 1) * chunk_size, total_kappa_combinations)
+                kappa_indices_chunk = kappa_indices[start_idx:end_idx]
+                tasks.append((task_idx, np.array([i]), ut, us_n, kappa1_flat, kappa2_flat, num_sim, data_slice, p_common, kappa_indices_chunk))
+                task_metadata[task_idx] = {'mean_indices': np.array([i]),
+                                           'p_common': p_common,
+                                           'kappa_indices': (start_idx, end_idx)}
+                task_idx += 1
+    with open('./learned_data/task_metadata.pkl', 'wb') as f:
+        pickle.dump(task_metadata, f)
+    initargs = (angle_gam_data_path, unif_fn_data_path)
+    print("Before creating multiprocessing pool")
+    # num_processes = int(os.environ['SLURM_CPUS_PER_TASK'])
+
+    log_folder = '/ceph/scratch/kdusterwald/slurm/logs/%j'
+    executor = submitit.AutoExecutor(folder=log_folder)
+    # slurm_array_parallelism tells the scheduler to only run at most 2 jobs at once. By default, this is several hundreds (no HPC default!)
```

Comment:
I guess the comment should be `at most 16 jobs`

Reply:
Yes, did not update :) btw not sure what the best limit is on jobs. I think less than 32 for sure, but 16 at a time seems more reasonable so as not to flood the cluster...